### PR TITLE
feat: skip app validating ops missing dependencies

### LINKED ISSUE #3532: [App validation workflow] Skip op validation if deps are missing
For ops whose dependencies are missing, validation should be skipped.

### PR DESCRIPTION
**Pull request by jost-s**
### Summary
Skip ops that have missing dependencies. If an op is awaiting dependencies to be fetched, it will be excluded from app validation.

closes #3532 

### TODO:
- [x] CHANGELOGs updated with appropriate info

### CODE REVIEW

File: crates/holochain/src/core/workflow/app_validation_workflow.rs [hunk 1]
```diff
@@ -50,53 +51,22 @@ mod unit_tests;
...
-
-    pub fn insert_missing_hash(&mut self, hash: AnyDhtHash) -> bool {
-        self.missing_hashes.insert(hash)
-    }
-
-    pub fn remove_missing_hash(&mut self, hash: &AnyDhtHash) -> bool {
-        self.missing_hashes.remove(hash)
-    }
-}
-
```
Comment (jost-s) [inline crates/holochain/src/core/workflow/app_validation_workflow.rs:83]: moved to the bottom of the file

File: crates/holochain/src/core/workflow/app_validation_workflow.rs [hunk 2]
```diff
@@ -125,6 +95,10 @@ async fn app_validation_workflow_inner(
 ) -> WorkflowResult<WorkComplete> {
     let db = workspace.dht_db.clone().into();
     let sorted_ops = validation_query::get_ops_to_app_validate(&db).await?;
+    // filter out ops that have missing dependencies
+    let sorted_ops = validation_dependencies
```
Comment (jost-s) [inline crates/holochain/src/core/workflow/app_validation_workflow.rs:99]: here ops missing dependencies are filtered out

File: crates/holochain/src/core/workflow/app_validation_workflow.rs [hunk 3]
```diff
@@ -156,16 +130,19 @@ async fn app_validation_workflow_inner(
                 });
 
                 // Validate this op
+                let dht_op_hash = op.to_hash();
```
Comment (jost-s) [inline crates/holochain/src/core/workflow/app_validation_workflow.rs:133]: for the map of missing dependency to DhtOp, the DhtOpHash is now passed all the way down to `run_validate_callback`.

File: crates/holochain/src/core/workflow/app_validation_workflow.rs [hunk 4]
```diff
@@ -445,35 +428,62 @@ async fn validate_op_outer(
...
-    R: RibosomeT,
-{
-    check_entry_def(op, &network.dna_hash(), conductor_handle)
+    validation_dependencies: Arc<Mutex<ValidationDependencies>>,
+) -> AppValidationOutcome<Outcome> {
+    check_entry_def(&op, &network.dna_hash(), conductor_handle)
         .await
         .map_err(AppValidationError::SysValidationError)?;
 
-    let zomes_to_invoke = match op {
```
Comment (jost-s) [inline crates/holochain/src/core/workflow/app_validation_workflow.rs:472]: this was inline and is now moved to a function

File: crates/holochain/src/core/workflow/app_validation_workflow/unit_tests.rs [hunk 5]
```diff
@@ -492,11 +519,177 @@ async fn validation_callback_prevent_multiple_identical_fetches() {
...
+                    }))))
+                }
+            }
+        }
+    });
+
+    let invocation = ValidateInvocation::new(zomes_to_invoke, &delete_action_op).unwrap();
+    let validation_dependencies = Arc::new(Mutex::new(ValidationDependencies::new()));
+
+    // hashes missing for delete dht op should be empty
```
Comment (jost-s) [inline crates/holochain/src/core/workflow/app_validation_workflow/unit_tests.rs:633]: testing here that hashes of missing deps are added and removed to the map

File: crates/holochain/src/core/workflow/app_validation_workflow.rs [hunk 6]
```diff
@@ -661,13 +658,14 @@ async fn store_record_zomes_to_invoke(
 
 async fn run_validation_callback(
     invocation: ValidateInvocation,
```
Comment (jost-s) [inline crates/holochain/src/core/workflow/app_validation_workflow.rs:660]: I had moved the invocation creation from the parent method `validate_op` here, but for some reason the test became flaky based on app entries' zome indexes not being found, so I reverted the change. It was cleaner with the invocation being instantiated in here... ☹️

**Review by ThetaSinner — APPROVED** [review]
Looks fantastic! Looking forward to having this change and not watching warning after warning as my ops try to re-validate and my room gets gradually warmer as my CPU heats up :D

**Pull request merged**